mail/views.py

```diff
@@ -108,22 +110,37 @@ def post(self, request, *args, **kargs):  # pylint: disable=unused-argument
             filter_on_email_optin=True
         )
         emails = get_all_query_matching_emails(search_obj)
+
+        automatic_email = None
         if request.data.get('send_automatic_emails'):
-            with transaction.atomic():
-                percolate_query = PercolateQuery.objects.create(query=search_obj.to_dict())
-                AutomaticEmail.objects.create(
-                    query=percolate_query,
-                    enabled=True,
-                    email_subject=email_subject,
-                    email_body=email_body,
-                    sender_name=sender_name,
-                )
-        MailgunClient.send_batch(
-            subject=email_subject,
-            body=email_body,
-            recipients=emails,
-            sender_name=sender_name,
-        )
+            automatic_email = add_automatic_email(
+                search_obj,
+                email_subject=request.data['email_subject'],
+                email_body=request.data['email_body'],
+                sender_name=sender_name,
+            )
+
+        try:
+            MailgunClient.send_batch(
+                subject=email_subject,
+                body=email_body,
+                recipients=emails,
+                sender_name=sender_name,
+            )
+
+            if automatic_email:
+                mark_emails_as_sent(automatic_email, emails)
+        except SendBatchException as send_batch_exception:
+            if automatic_email:
+                failure_emails = set()
+                for recipients, _ in send_batch_exception.exception_pairs:
+                    failure_emails |= set(recipients)
+                success_emails = set(emails).difference(failure_emails)
```

Comment:
can you change `SendBatchException` so that you just need to refer to `send_batch_exception.failed_recipient_emails` here instead of doing all of this calculation in the `except` block? could be a property or function. this would be a lot more readable

```python
except SendBatchException as send_batch_exception:
    if automatic_email:
        success_emails = set(emails).difference(send_batch_exception.failed_recipient_emails)
        mark_emails_as_sent(automatic_email, success_emails)
```